### PR TITLE
loop wait rollout of loggging audit event

### PR DESCRIPTION
**Pull request by w13915984028**
loop wait rollout of loggging audit event
Signed-off-by: Jian Wang <w13915984028@gmail.com>

**IMPORTANT: Please do not create a Pull Request without creating an issue first.**

**Problem:**

The depolyment time of `logging audit event` varies in different environment, simply wait 50 seconds is not enough.

**Solution:**

Loop wait for a maximum of time, avoid simple failure.

**Related Issue:**
https://github.com/harvester/harvester/issues/2931

**Test plan:**

upgrade from v1.0.3 to v1.1.0 (e.g. v1.1.0-rc3), check if the upgrade-manifest POD does not fail due to failure: wait rollout logging audit event

### CODE REVIEW

**Comment by w13915984028** [comment]
local test:

```
In any harvester running environment:

harv41:~ # curl https://github.com/w13915984028/harvester-develop-summary/tree/main/test-code/loop_wait_rollout/loop_wait_rollout.sh

harv41:~ # ./wait_loop.sh 
Waiting for deployment "rancher-logging" rollout to finish: 0 of 1 updated replicas are available...
deployment "rancher-logging" successfully rolled out
daemon set "rancher-logging-root-fluentbit" successfully rolled out
daemon set "rancher-logging-rke2-journald-aggregator" successfully rolled out
daemon set "rancher-logging-kube-audit-fluentbit" successfully rolled out
success to wait rollout logging audit
partitioned roll out complete: 1 new pods have been updated...
success to wait rollout event
harv41:~ # 

harv41:~ # 
```

change any waiting object to be invalid, then: e.g. `statefulset harvester-default-event-tailer1`, the script will loop

```
harv41:~ # ./wait_loop.sh 
deployment "rancher-logging" successfully rolled out
daemon set "rancher-logging-root-fluentbit" successfully rolled out
daemon set "rancher-logging-rke2-journald-aggregator" successfully rolled out
daemon set "rancher-logging-kube-audit-fluentbit" successfully rolled out
success to wait rollout logging audit
Error from server (NotFound): statefulsets.apps "harvester-default-event-tailer1" not found
continue waiting rollout statefulset harvester-default-event-tailer1, 1
Error from server (NotFound): statefulsets.apps "harvester-default-event-tailer1" not found
continue waiting rollout statefulset harvester-default-event-tailer1, 2
Error from server (NotFound): statefulsets.apps "harvester-default-event-tailer1" not found
continue waiting rollout statefulset harvester-default-event-tailer1, 3
Error from server (NotFound): statefulsets.apps "harvester-default-event-tailer1" not found
continue waiting rollout statefulset harvester-default-event-tailer1, 4
^C

```

**Comment by w13915984028** [comment]
Validated in single-node cluster, the `loop_wait` runs as expected.

```
The current version is v1.0.3, will check Logging Event Audit upgrade manifest option
Logging Event Audit: start to upgrade manifest
..
Apply resource file of logging and audit
namespace/cattle-logging-system created
managedchart.management.cattle.io/rancher-logging-crd created
managedchart.management.cattle.io/rancher-logging created
Wait for rollout of logging and audit
deployment "rancher-logging" successfully rolled out
daemon set "rancher-logging-root-fluentbit" successfully rolled out
daemon set "rancher-logging-rke2-journald-aggregator" successfully rolled out
daemon set "rancher-logging-kube-audit-fluentbit" successfully rolled out
success to wait rollout logging audit
..
add event resource from /usr/local/share/migrations/managed_charts/event-v1.0.3.yaml to manifest rancher-logging.yaml
Apply resource file of event
eventtailer.logging-extensions.banzaicloud.io/harvester-default created
Wait for rollout of event
partitioned roll out complete: 1 new pods have been updated...
success to wait rollout event
..
Logging Event Audit: finish upgrading manifest
Applying extra manifests

```

pod `apply-manifests` runs once
```
harv41:~ # kk get pods -A | grep upgrade
cattle-system               apply-hvst-upgrade-6ccvj-prepare-on-harv41-with-9e26d5d34-zkh8s   0/1     Completed     0          15m
cattle-system               apply-system-agent-upgrader-on-harv41-with-7640d5ee2bf0ab-nn42c   0/1     Completed     0          9m22s
cattle-system               system-upgrade-controller-7b8d94c7f5-fcctk                        1/1     Running       0          9m15s
harvester-system            hvst-upgrade-6ccvj-apply-manifests-q85lx                          0/1     Completed     0          10m
harvester-system            hvst-upgrade-6ccvj-single-node-upgrade-harv41-bflbp               1/1     Running       0          2m22s
```

File: package/upgrade/upgrade_manifests.sh [hunk 1]
```diff
@@ -453,27 +531,16 @@ upgrade_logging_event_audit() {
     # reuse frame work to generate yaml file
     upgrade_managed_chart_from_version $UPGRADE_PREVIOUS_VERSION rancher-logging rancher-logging.yaml
 
-    echo "Apply resource file"
+    echo "Apply resource file of logging and audit"
 
     kubectl apply -f ./rancher-logging.yaml
 
     # wait for managedchart to be applied
     sleep 50
```
Comment (Vicente-Cheng) [inline package/upgrade/upgrade_manifests.sh:539]: just curious, If we have the retry mechanism on the following. Do we still need to sleep 50 seconds here?

File: package/upgrade/upgrade_manifests.sh [hunk 2]
```diff
@@ -485,17 +552,16 @@ upgrade_logging_event_audit() {
     # rancher-logging_event-extension is reusing chart rancher-logging, but as an extension for event
     upgrade_managed_chart_from_version $UPGRADE_PREVIOUS_VERSION rancher-logging_event-extension rancher-logging.yaml
 
-    echo "Apply resource file"
+    echo "Apply resource file of event"
 
     kubectl apply -f ./rancher-logging.yaml
 
     # wait few seconds
     sleep 20
```
Comment (Vicente-Cheng) [inline package/upgrade/upgrade_manifests.sh:560]: ditto, do we need to sleep 20 seconds here?

**Review by Vicente-Cheng — COMMENTED** [review]
some nits, others LGTM.

File: package/upgrade/upgrade_manifests.sh (hunk 1 above)
Comment (w13915984028) [reply package/upgrade/upgrade_manifests.sh:539]: The consideration:
(1) The first sleep will help reduce looping message. From practice, the `managedchart` may take 30 seconds ~3 minutes to be applied. The loop waiting interval is 10s. 
(2) The fix tries to change exising code as few as possible.

File: package/upgrade/upgrade_manifests.sh (hunk 2 above)
Comment (w13915984028) [reply package/upgrade/upgrade_manifests.sh:560]: similar consideration

**Review by starbops — APPROVED** [review]
LGTM and tested on 4-node cluster, thanks!

```
The current version is v1.0.3, will check Logging Event Audit upgrade manifest option
Logging Event Audit: start to upgrade manifest
+ CHART_NAME=rancher-logging
+ CHART_MANIFEST=rancher-logging.yaml
+ case $CHART_NAME in
+ create_logging_audit
+ RESOURCE_FILE=/usr/local/share/migrations/managed_charts/logging-audit-v1.0.3.yaml
+ '[' -e /usr/local/share/migrations/managed_charts/logging-audit-v1.0.3.yaml ']'
+ echo 'add logging audit resource from /usr/local/share/migrations/managed_charts/logging-audit-v1.0.3.yaml to manifest rancher-logging.yaml'
+ cat /usr/local/share/migrations/managed_charts/logging-audit-v1.0.3.yaml
add logging audit resource from /usr/local/share/migrations/managed_charts/logging-audit-v1.0.3.yaml to manifest rancher-logging.yaml
Apply resource file of logging and audit
namespace/cattle-logging-system created
managedchart.management.cattle.io/rancher-logging-crd created
managedchart.management.cattle.io/rancher-logging created
Wait for rollout of logging and audit
Error from server (NotFound): deployments.apps "rancher-logging" not found
continue waiting rollout deployment rancher-logging, 1
Error from server (NotFound): deployments.apps "rancher-logging" not found
continue waiting rollout deployment rancher-logging, 2
deployment "rancher-logging" successfully rolled out
Error from server (NotFound): daemonsets.apps "rancher-logging-root-fluentbit" not found
continue waiting rollout daemonset rancher-logging-root-fluentbit, 3
deployment "rancher-logging" successfully rolled out
Error from server (NotFound): daemonsets.apps "rancher-logging-root-fluentbit" not found
continue waiting rollout daemonset rancher-logging-root-fluentbit, 4
deployment "rancher-logging" successfully rolled out
daemon set "rancher-logging-root-fluentbit" successfully rolled out
daemon set "rancher-logging-rke2-journald-aggregator" successfully rolled out
Waiting for daemon set "rancher-logging-kube-audit-fluentbit" rollout to finish: 0 of 4 updated pods are available...
Waiting for daemon set "rancher-logging-kube-audit-fluentbit" rollout to finish: 1 of 4 updated pods are available...
Waiting for daemon set "rancher-logging-kube-audit-fluentbit" rollout to finish: 2 of 4 updated pods are available...
Waiting for daemon set "rancher-logging-kube-audit-fluentbit" rollout to finish: 3 of 4 updated pods are available...
daemon set "rancher-logging-kube-audit-fluentbit" successfully rolled out
success to wait rollout logging audit
+ CHART_NAME=rancher-logging_event-extension
+ CHART_MANIFEST=rancher-logging.yaml
+ case $CHART_NAME in
+ create_event
+ RESOURCE_FILE=/usr/local/share/migrations/managed_charts/event-v1.0.3.yaml
+ '[' -e /usr/local/share/migrations/managed_charts/event-v1.0.3.yaml ']'
add event resource from /usr/local/share/migrations/managed_charts/event-v1.0.3.yaml to manifest rancher-logging.yaml
+ echo 'add event resource from /usr/local/share/migrations/managed_charts/event-v1.0.3.yaml to manifest rancher-logging.yaml'
+ cat /usr/local/share/migrations/managed_charts/event-v1.0.3.yaml
Apply resource file of event
eventtailer.logging-extensions.banzaicloud.io/harvester-default created
Wait for rollout of event
partitioned roll out complete: 1 new pods have been updated...
success to wait rollout event
Logging Event Audit: finish upgrading manifest
```

**Review by bk201 — APPROVED** [review]
LGTM

File: package/upgrade/upgrade_manifests.sh (hunk 1 above)
Comment (Vicente-Cheng) [reply package/upgrade/upgrade_manifests.sh:539]: OK, thanks for the explanation!

**Review by Vicente-Cheng — APPROVED** [review]
LGTM, thanks!

**Pull request merged**